src/utils/HTTPTransport.ts

```diff
@@ -0,0 +1,107 @@
+enum METHOD {
+  GET = 'GET',
+  POST = 'POST',
+  PUT = 'PUT',
+  PATCH = 'PATCH',
+  DELETE = 'DELETE',
+}
+
+interface IHeaders {
+  [k: string]: string;
+}
+
+interface IRequestBody {
+  [k: string]: unknown;
+}
+
+type Options = {
+  method: METHOD;
+  headers?: IHeaders;
+  timeout?: number;
+  data?: IRequestBody;
+};
+
+type OptionsWithoutMethod = Omit<Options, 'method'>;
+
+function parseOptions(options: Options) {
+  const { method } = options;
+  let { headers, data } = options;
+  if (!headers || Object.keys(headers).length === 0) {
+    headers = undefined;
+  }
+  if (!data || Object.keys(data).length === 0) {
+    data = undefined;
+  }
+  return { method, headers, data };
+}
+
+function queryStringify(data: IRequestBody) {
+  return Object.entries(data)
+    .map(([key, value]) => `${key}=${encodeURIComponent(String(value))}`)
+    .join('&');
+}
+export class HTTPTransport {
+  get(url: string, options: OptionsWithoutMethod = {}): Promise<XMLHttpRequest> {
```

Comment:
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod = (url: string, options?: OptionsWithoutMethod) => Promise<XMLHttpRequest>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

Reply:
Из документации typescript:

[Arrow Functions](https://www.typescriptlang.org/docs/handbook/2/classes.html#arrow-functions)

- This will use more memory, because each class instance will have its own copy of each function defined this way
- You can’t use `super.<method-name>` in a derived class, because there’s no entry in the prototype chain to fetch the base class method from

А использование стандартных методов класса (а не свойств в виде стрелочных функций) требует явного задания типов параметров и возвращаемого значения, насколько я понял из своих изысканий.